### PR TITLE
feat: add `iter->status()` check for loop iterators

### LINKED ISSUE #2200: Check iterator status after quit loop
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Motivation

From [rocksdb doc](https://github.com/facebook/rocksdb/wiki/Iterator#error-handling), need handle error when iterating.   Like bellow:
```
for (it->Seek("hello"); it->Valid(); it->Next()) {
  // Do something with it->key() and it->value().
}
if (!it->status().ok()) {
  // Handle error. it->status().ToString() contains error message.
}
```

However, the code in kvrocks does not do this check. 
For instance: 
https://github.com/apache/kvrocks/blob/unstable/src/types/redis_hash.cc#L344

```
rocksdb::Status Hash::GetAll(const Slice &user_key, std::vector<FieldValue> *field_values, HashFetchType type) {
  field_values->clear();

  std::string ns_key = AppendNamespacePrefix(user_key);
  HashMetadata metadata(false);
  LatestSnapShot ss(storage_);
  rocksdb::Status s = GetMetadata(GetOptions{.snapshot = ss.GetSnapShot()}, ns_key, &metadata);
  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;

  std::string prefix_key = InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode();
  std::string next_version_prefix_key =
      InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode();

  rocksdb::ReadOptions read_options = storage_->DefaultScanOptions();
  read_options.snapshot = ss.GetSnapShot();
  rocksdb::Slice upper_bound(next_version_prefix_key);
  read_options.iterate_upper_bound = &upper_bound;

  auto iter = util::UniqueIterator(storage_, read_options);
  for (iter->Seek(prefix_key); iter->Valid() && iter->key().starts_with(prefix_key); iter->Next()) {
    if (type == HashFetchType::kOnlyKey) {
      InternalKey ikey(iter->key(), storage_->IsSlotIdEncoded());
      field_values->emplace_back(ikey.GetSubKey().ToString(), "");
    } else if (type == HashFetchType::kOnlyValue) {
      field_values->emplace_back("", iter->value().ToString());
    } else {
      InternalKey ikey(iter->key(), storage_->IsSlotIdEncoded());
      field_values->emplace_back(ikey.GetSubKey().ToString(), iter->value().ToString());
    }
  }
  return rocksdb::Status::OK();
}
```

### Solution

check iterator status.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by LindaSummer**
# Issue

close #2200

# Proposed Changes
- add iterator status error handler after iteration.

# Comment
1. Some iterators checked `Valid()` after loop so I didn't add status validation.
    Here is [rocksdb's document](https://github.com/facebook/rocksdb/wiki/Iterator#error-handling).
> If there is no error, the status is `Status::OK()`. If the status is not OK, the iterator will be invalidated too. In another word, if `Iterator::Valid()` is true, `status()` is guaranteed to be `OK()` so it's safe to proceed other operations without checking `status()`.

2. Please correct me if I need add some more logic of error handler, or the error message I generated is not suitable.
Thanks very much!

### CODE REVIEW

File: src/cluster/slot_migrate.cc [hunk 1]
```diff
@@ -382,6 +382,12 @@ Status SlotMigrator::sendSnapshotByCmd() {
     }
   }
 
+  if (auto s = iter->status(); !s.ok()) {
+    auto err_str = s.ToString();
+    LOG(ERROR) << "[migrate] Failed to iterate keys of slot " << slot << ": " << err_str;
+    return {Status::NotOK, fmt::format("failed to iterate keys of slot {}: {}", slot, s.ToString())};
```
Comment (git-hulk) [inline src/cluster/slot_migrate.cc]: Can also use err_str here.

**Comment by git-hulk** [comment]
LGTM, one comment inline.

File: src/cluster/slot_migrate.cc (hunk 1 above)
Comment (LindaSummer) [reply src/cluster/slot_migrate.cc]: Hi @git-hulk ,

Thanks for your suggestion!
I have replaced `s.ToString()` with `err_str`.

Best Regards,
Edward

**Review by git-hulk — APPROVED** [review]

**Comment by LindaSummer** [comment]
Hi @git-hulk ,

I find it seems hard to create unit tests for the error handling code and pass sonarcloud coverage.
Could you give me some suggestions?

Best Regards,
Edward

**Comment by git-hulk** [comment]
@LindaSummer That’s fine for this PR, no need to add the extra tests.

**Comment by LindaSummer** [comment]
Hi @git-hulk ,

Got it! 😊
Thanks very much for your help and patience!

Best Regards,
Edward

**Pull request merged**